File: 0001~0500/0092. Reverse Linked List II/main.py
"""
Source: https://leetcode.com/problems/reverse-linked-list-ii/
Date: 2023/2/10
Skill:
Runtime: 30 ms, faster than 85.75%
Memory Usage: 14 MB, less than 44.46%
Time complexity:
Space complexity:
Constraints:
    1 <= n <= 500
    1 <= left <= right <= n
"""

from typing import List, Optional
from collections import defaultdict

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right: return head
        dummy_head = ListNode(next=head)
        cur = dummy_head
        for _ in range(left-1):
            cur = cur.next
        first_tail, second_tail, prev, cur, next = cur, cur.next, None, cur.next, cur.next.next
        for _ in range(right-left):
            cur.next = prev
            prev = cur
            cur = next
            next = cur.next

        cur.next = prev
        first_tail.next = cur
        second_tail.next = next
        return dummy_head.next


if __name__ == "__main__":
    s = Solution

File: 0001~0500/0092. Reverse Linked List II/main.cpp
/**
 * Source: https://rb.gy/3xpqp
 * Date: 2023/9/7
 * Skill:
 * Runtime: 4 ms, faster than 38.20% of C++ online submissions
 * Memory Usage: 7.5 MB, less than 12.22% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long

typedef pair<int, int> pairs;



struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (left == right) return head;
        ListNode* pseudo_head = new ListNode(0, head);
        ListNode* l = pseudo_head;
        for (int i = 1; i < left; i++)
            l = l->next;
        ListNode* cur = l->next, *next = cur->next;
        for (int i = 1; i <= right - left; i++) {
            ListNode* tmp = next->next;
            next->next = cur;
            cur = next;
            next = tmp;
        }
        l->next->next = next;
        l->next = cur;
        return pseudo_head->next;
    }
};


static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

int main() {
//    Solution s;
    string s = "1";
    cout << s[0] - '0';
}